File: Binary_Search_Tree/235.Lowest_Common_Ancestor_of_a_Binary_Search_Tree/C++/235.Lowest_Common_Ancestor_of_a_Binary_Search_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode * lowestCommonAncestor(TreeNode * root, TreeNode * p, TreeNode * q) 
    {
        //========================================//
        // Input type:                            //
        //   - TreeNode * (root tree-node)        //
        //   - TreeNode * (p tree-node)           //
        //   - TreeNode * (q tree-node)           //
        // Return type:                           //
        //   - TreeNode * (result root tree-node) //
        //========================================//

        /*Initialize*/
        ///// Record maximun value of child-nodes /////
        int record_max_val = ((p->val) > (q->val)) ? (p->val) : (q->val);

        ///// Record minimun value of child-nodes /////
        int record_min_val = ((p->val) < (q->val)) ? (p->val) : (q->val);


        /*Binary-search based loop traversal*/
        while (root != nullptr) //Binary-search
        {
            ///// Check if the current root-node value matched conditions or not /////
            if ((root->val) < record_min_val) { root = (root->right); } //Keep updating, traversing

            else if ((root->val) > record_max_val) { root = (root->left); } //Keep updating, traversing

            else { return root; }

        } //Binary-search

        return nullptr;
    }
};

File: Binary_Search_Tree/235.Lowest_Common_Ancestor_of_a_Binary_Search_Tree/Python/235.Lowest_Common_Ancestor_of_a_Binary_Search_Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        #======================================#
        # Binary-search based traversal method #
        #======================================#

        ############
        #Initialize
        ##### Record maximun root-node value #####
        record_max_val = (p.val if (p.val > q.val) else q.val)

        ##### Record minimun root-node value #####
        record_min_val = (p.val if (p.val < q.val) else q.val)


        ####################################
        #Binary-search based loop traversal
        while (root):

            ##### Check if the current root-node value matched conditions or not #####
            if (root.val > record_max_val):
                root = root.left #Keep updating/traversing

            elif (root.val < record_min_val):
                root = root.right #Keep updating/traversing

            else:
                return root

        return None
